File: Snake_PyQt/Scene/GameScene.py
from Scene.Scene import Scene
from Director import Director
from Config import Config
from Scene.StartScene import StartScene
from PyQt5.QtWidgets import QMessageBox, QInputDialog
from PyQt5.QtGui import QPalette, QColor, QPainter, QBrush, QFont
from PyQt5.QtCore import Qt, QTimer
from Sprite.Snake import Snake
from Sprite.Food import Food


#游戏主场景
class GameScene(Scene):
    def __init__(self, parentWindow):
        super().__init__(parentWindow)

    #初始化场景
    #传入从0开始，1为递增量的难度系数
    def init(self, difficulty):
        #获取宽高方向地图线的数目
        self.spaceNum = 20

        #获取宽和高方向的间隔
        self.widthSpace = int(self.width() / self.spaceNum)
        self.heightSpace = int(self.height() / self.spaceNum)

        #设定FPS值
        #平常的按阶梯递增的难度
        if difficulty == 3:
            self.fps = 1000
        #光速难度
        else:
            self.fps = difficulty * 5 + 5

        #初始化蛇
        self.snake = Snake([5, 5], 2, 3)
        #初始化食物
        self.food = Food()
        self.food.bodyGrowth = difficulty + 1
        self.food.updatePos(-1, self.spaceNum, -1, self.spaceNum,
                            self.snake.pos)

        #初始化当前分数
        self.score = 0
        #设置分数增长
        self.scoreGrowth = difficulty + 1

        #设置背景颜色
        self.setAutoFillBackground(True)
        self.setPalette(QPalette(QColor(255, 255, 255)))

        #绘制画面定时器
        self.timer = QTimer(self)
        self.timer.timeout.connect(self.gameCycle)
        self.timer.setInterval(1000 / self.fps)
        self.timer.start()

    #重写按键事件
    def keyPressEvent(self, event):
        if not event.isAutoRepeat():
            if event.key() == Qt.Key_Up and abs(
                    self.snake.pos[0][1] - self.snake.pos[1][1]) != 1:
                self.snake.dir = 0
            elif event.key() == Qt.Key_Down and abs(
                    self.snake.pos[0][1] - self.snake.pos[1][1]) != 1:
                self.snake.dir = 1
            elif event.key() == Qt.Key_Left and abs(
                    self.snake.pos[0][0] - self.snake.pos[1][0]) != 1:
                self.snake.dir = 2
            elif event.key() == Qt.Key_Right and abs(
                    self.snake.pos[0][0] - self.snake.pos[1][0]) != 1:
                self.snake.dir = 3
            elif event.key() == Qt.Key_Space:
                self.timer.setInterval(self.timer.interval() / 3)
        return super().keyPressEvent(event)

    #按键释放的时候的事件
    def keyReleaseEvent(self, event):
        if not event.isAutoRepeat():
            if event.key() == Qt.Key_Space:
                self.timer.setInterval(self.timer.interval() * 3)
        return super().keyReleaseEvent(event)

    #游戏循环
    def gameCycle(self):
        #蛇的移动
        self.snake.move()

        #如果吃到了食物
        if self.snake.eatFood(self.food):
            #刷新分数
            self.score += self.scoreGrowth
            #刷新食物坐标
            self.food.updatePos(-1, self.spaceNum, -1, self.spaceNum,
                                self.snake.pos)

        #判断游戏是否结束
        if self.snake.isDead(0, self.spaceNum - 1, 0, self.spaceNum - 1):
            self.gameEnd()

        #更新界面，触发paintEvent事件重绘界面
        self.update()

    #绘制事件
    def paintEvent(self, event):
        painter = QPainter(self)

        #绘制背景线
        for i in range(self.spaceNum):
            painter.drawLine(i * self.widthSpace, 0, i * self.widthSpace,
                             self.height())
        for i in range(self.spaceNum):
            painter.drawLine(0, i * self.heightSpace, self.width(),
                             i * self.heightSpace)

        #绘制蛇身
        painter.setBrush(QBrush(QColor(0, 0, 255)))
        for i, j in self.snake.pos:
            painter.drawRect(i * self.widthSpace, j * self.heightSpace,
                             self.widthSpace, self.heightSpace)

        #绘制食物
        painter.setBrush(QBrush(QColor(255, 0, 0)))
        painter.drawRect(self.food.x * self.widthSpace,
                         self.food.y * self.heightSpace, self.widthSpace,
                         self.heightSpace)

        #绘制分数
        painter.setPen(Qt.red)
        painter.setFont(QFont("微软雅黑", 13, 5))
        painter.drawText(0, 0, self.width(), 25, Qt.AlignCenter,
                         "分数：%d" % self.score)

        painter.end()
        return super().paintEvent(event)

    #游戏结束
    def gameEnd(self):
        #定时器停止
        self.timer.stop()

        #判断是否是最高分
        #如果是的话，则保存最高分并且录入保持者名字
        if self.score > Config().highestScore:
            Config().highestScore = self.score
            highestScorePlayer = ""
            #必须有纪录保持者
            while highestScorePlayer == "":
                highestScorePlayer = QInputDialog.getText(
                    self, self.tr("最高分！"),
                    self.tr("恭喜你的分数 %d 成为最高分，请留下你的尊姓大名。") % self.score)[0]
            #录入最高分纪录保持者
            Config().highestScorePlayer = highestScorePlayer
        #判断是否要再来一局
        #如果不要的话回到开始场景
        if QMessageBox.information(
                self, self.tr("游戏结束"),
                self.tr("您的得分是：%d，是否要以当前难度再来一局？") % self.score,
                QMessageBox.Yes, QMessageBox.No) == QMessageBox.No:
            startScene = StartScene(Director().window)
            Director().nowScene = startScene
            startScene.init()
            startScene.show()
            self.deleteLater()
        #否则将游戏的各项数据设置复原
        else:
            #重置蛇和食物
            self.snake = Snake([5, 5], 2, 3)
            self.food.updatePos(-1, self.spaceNum, -1, self.spaceNum,
                                self.snake.pos)
            #重置分数
            self.score = 0
            #重置绘制定时器
            self.timer.setInterval(1000 / self.fps)
            self.timer.start()

File: snake_qt5/GameScene.cpp
#include "GameScene.h"
#include "Director.h"
#include "Config.h"
#include "StartScene.h"
#include "QPainter"
#include "QTimer"
#include "QKeyEvent"
#include "QMessageBox"
#include "QInputDialog"

GameScene::GameScene(Window *parent) :Scene(parent)
{

}

GameScene::~GameScene()
{
}

void GameScene::init(int difficulty)
{
	//初始化游戏数据参数等
	//初始化地图线条数目及其间隔
	lineNum = 25;
	interval = static_cast<double>(width()) / (lineNum + 1);

	//初始化游戏帧数
	fps = 1000 / (200 - difficulty * 50);

	//初始化蛇
	snake = Snake({ 2,2 }, 2, 3);
	//初始化食物
	food.setBodyGrowth(difficulty + 1);
	food.updatePos(0, lineNum, 0, lineNum, snake.getPos());

	//初始化当前分数
	nowScore = 0;
	//初始化分数增加数目
	scoreGrowth = difficulty + 1;

	//设置背景
	setAutoFillBackground(true);
	setPalette(QColor(255, 255, 255));

	//初始化画面更新定时器并开始更新界面
	paintTimer = new QTimer(this);
	connect(paintTimer, &QTimer::timeout, this, &GameScene::updateSnake);
	paintTimer->setInterval(1000 / fps);
	paintTimer->start();
}

void GameScene::updateSnake()
{
	//蛇移动
	snake.move();

	//判定是否吃到食物
	if (snake.eatFood(food))
	{
		//分数刷新
		nowScore += scoreGrowth;

		//更新食物坐标
		food.updatePos(0, lineNum, 0, lineNum, snake.getPos());
	}

	//判定死亡
	if (snake.isDead(0, lineNum, 0, lineNum))
	{
		//停止更新画面
		paintTimer->stop();

		//判定分数是否为最高分
		if (nowScore > Config::getInstance()->getHighestScore())
		{
			//把最高分和最高分保持者保存进Config类中
			Config::getInstance()->setHighestScore(nowScore);
			Config::getInstance()->setHighestScorePlayer(QInputDialog::getText(this, tr(u8"最高分"), tr(u8"恭喜这位少侠！你的分数是：%1，超过了以往所有的记录，为最高分！请在这里记录下你的大名。").arg(nowScore)));
		}

		//是否再来一局
		if (QMessageBox::information(this, tr(u8"游戏结束"), tr(u8"少年，你挂了！你的最终分数是：%1，是否要按照当前速度再来一局？").arg(nowScore), QMessageBox::Yes, QMessageBox::No) == QMessageBox::No)
		{
			StartScene *startScene = new StartScene(Director::getInstance()->getWindow());
			Director::getInstance()->setNowScene(startScene);
			startScene->init();
			startScene->show();
			deleteLater();
			return;
		}
		//再来一局则重新初始化游戏数据等内容
		//初始化分数
		nowScore = 0;

		//初始蛇
		snake = Snake({ 2,2 }, 2, 3);

		//更新食物坐标
		food.updatePos(0, lineNum, 0, lineNum, snake.getPos());

		//开始更新画面
		paintTimer->setInterval(1000 / fps);
		paintTimer->start();

		return;
	}

	//绘制画面
	update();
}

void GameScene::paintEvent(QPaintEvent *event)
{
	QPainter *painter = new  QPainter(this);

	//绘制背景线
	for (int i = 1; i <= lineNum; i++)
	{
		painter->drawLine(0, i*interval, width(), i*interval);
		painter->drawLine(i*interval, 0, i*interval, height());
	}

	//绘制蛇身
	painter->setPen(QPen(Qt::blue));
	painter->setBrush(QBrush(Qt::blue));
	for (QPair<int, int> pos : snake.getPos())
	{
		painter->drawRect(pos.first*interval, pos.second*interval, interval, interval);
	}

	//绘制食物
	painter->setPen(QPen(Qt::red));
	painter->setBrush(QBrush(Qt::red));
	painter->drawRect(food.getPos().first*interval, food.getPos().second*interval, interval, interval);

	//绘制分数
	painter->setFont(QFont(u8"微软雅黑", 12));
	painter->drawText(0, 0, width(), 30, Qt::AlignCenter, QString(tr(u8"分数：%1").arg(nowScore)));

	painter->end();
}

void GameScene::keyPressEvent(QKeyEvent *event)
{
	//按键控制
	//方向控制
	auto snakePos = snake.getPos();
	if (event->key() == Qt::Key_Up && snakePos[0].second - snakePos[1].second != 1)
	{
		snake.setNowDir(0);
	}
	else if (event->key() == Qt::Key_Down && snakePos[0].second - snakePos[1].second != -1)
	{
		snake.setNowDir(1);
	}
	else if (event->key() == Qt::Key_Left && snakePos[0].first - snakePos[1].first != 1)
	{
		snake.setNowDir(2);
	}
	else if (event->key() == Qt::Key_Right && snakePos[0].first - snakePos[1].first != -1)
	{
		snake.setNowDir(3);
	}
	//按住空格加速
	else if (event->key() == Qt::Key_Space && !event->isAutoRepeat())
	{
		paintTimer->setInterval(paintTimer->interval() / 3);
	}
}

void GameScene::keyReleaseEvent(QKeyEvent * event)
{
	//释放空格回到原来的速度
	if (event->key() == Qt::Key_Space && !event->isAutoRepeat())
	{
		paintTimer->setInterval(paintTimer->interval() * 3);
	}
}